In the `where` clause of criteria, Using ValueObject of record annotated with `@DataType` will occur `DOMA1007`.

**Describe the bug**
In the `where` clause of criteria,
Using ValueObject of record annotated with `@DataType` will occur `DOMA1007`.
Also, when annotated with `@Domain`, it works as expected.

**Expected behavior**
I expect that `DOMA1007` will not occur and will work.

**To Reproduce**
Will reproduce on [Repository by forking getting-started](https://github.com/mazeneko/doma-data-type-issue).
See the commit history for details.

**Screenshots**
```
PS C:\doma-data-type-issue> ./gradlew run

> Task :java-17:run FAILED
Exception in thread "main" org.seasar.doma.internal.jdbc.scalar.ScalarException: [DOMA1007] The wrapper class for the type "boilerplate.java17.entity.Employee$Name" is not found.
        at org.seasar.doma.internal.jdbc.scalar.Scalars.wrap(Scalars.java:73)
        at org.seasar.doma.jdbc.criteria.context.Operand$Param.createInParameter(Operand.java:29)
        at org.seasar.doma.jdbc.criteria.query.BuilderSupport.param(BuilderSupport.java:195)
        at org.seasar.doma.jdbc.criteria.query.BuilderSupport$OperandVisitor.visit(BuilderSupport.java:210)
        at org.seasar.doma.jdbc.criteria.query.BuilderSupport$OperandVisitor.visit(BuilderSupport.java:207)
        at org.seasar.doma.jdbc.criteria.context.Operand$Param.accept(Operand.java:40)
        at org.seasar.doma.jdbc.criteria.query.BuilderSupport$CriterionVisitor.comparison(BuilderSupport.java:371)
        at org.seasar.doma.jdbc.criteria.query.BuilderSupport$CriterionVisitor.visit(BuilderSupport.java:230)
        at org.seasar.doma.jdbc.criteria.context.Criterion$Eq.accept(Criterion.java:24)
        at org.seasar.doma.jdbc.criteria.query.BuilderSupport.visitCriterion(BuilderSupport.java:204)
        at org.seasar.doma.jdbc.criteria.query.SelectBuilder.visitCriterion(SelectBuilder.java:242)
        at org.seasar.doma.jdbc.criteria.query.SelectBuilder.where(SelectBuilder.java:143)
        at org.seasar.doma.jdbc.criteria.query.SelectBuilder.interpret(SelectBuilder.java:72)
        at org.seasar.doma.jdbc.criteria.query.SelectBuilder.build(SelectBuilder.java:64)
        at org.seasar.doma.jdbc.criteria.statement.EntityqlSelectTerminal.createCommand(EntityqlSelectTerminal.java:50)
        at org.seasar.doma.jdbc.criteria.statement.EntityqlSelectStarting.createCommand(EntityqlSelectStarting.java:161)
        at org.seasar.doma.jdbc.criteria.statement.AbstractStatement.execute(AbstractStatement.java:36)
        at org.seasar.doma.jdbc.criteria.statement.Listable.fetch(Listable.java:23)
        at boilerplate.java17.repository.EmployeeRepository.selectByName(EmployeeRepository.java:31)
        at boilerplate.java17.Main.lambda$main$0(Main.java:26)
        at org.seasar.doma.jdbc.tx.LocalTransactionManager.lambda$toSupplier$0(LocalTransactionManager.java:242)
        at org.seasar.doma.jdbc.tx.LocalTransactionManager.executeInTransaction(LocalTransactionManager.java:262)
        at org.seasar.doma.jdbc.tx.LocalTransactionManager.requiredInternal(LocalTransactionManager.java:128)
        at org.seasar.doma.jdbc.tx.LocalTransactionManager.required(LocalTransactionManager.java:88)
        at boilerplate.java17.Main.main(Main.java:23)

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':java-17:run'.
> Process 'command 'C:\Program Files\Eclipse Adoptium\jdk-17.0.9.9-hotspot\bin\java.exe'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 3s
3 actionable tasks: 2 executed, 1 up-to-date
```

**Environment (please complete the following information):**
- Doma version or git rev: 3.0.1
- Build tool (ie. output of `mvnw --version` or `gradlew --version`): Gradle 8.10.2
- Database version: H2 2.3.232

--- a/doma-core/src/main/java/org/seasar/doma/internal/jdbc/scalar/Scalars.java
+++ b/doma-core/src/main/java/org/seasar/doma/internal/jdbc/scalar/Scalars.java
@@ -16,6 +16,7 @@ import java.time.LocalDate;
 import java.time.LocalDateTime;
 import java.time.LocalTime;
 import java.util.function.Supplier;
+import org.seasar.doma.DataType;
 import org.seasar.doma.Domain;
 import org.seasar.doma.jdbc.ClassHelper;
 import org.seasar.doma.jdbc.domain.DomainType;
@@ -255,7 +256,8 @@ public final class Scalars {
   protected static <BASIC, DOMAIN> Supplier<Scalar<?, ?>> wrapDomainObject(
       Object value, Class<DOMAIN> valueClass, boolean optional, ClassHelper classHelper) {
     DomainType<BASIC, DOMAIN> domainType;
-    if (valueClass.isAnnotationPresent(Domain.class)) {
+    if (valueClass.isAnnotationPresent(Domain.class)
+        || valueClass.isAnnotationPresent(DataType.class)) {
       domainType = DomainTypeFactory.getDomainType(valueClass, classHelper);
     } else {
       domainType = DomainTypeFactory.getExternalDomainType(valueClass, classHelper);

--- a/integration-test-java/src/main/java/org/seasar/doma/it/criteria/Avenue.java
+++ b/integration-test-java/src/main/java/org/seasar/doma/it/criteria/Avenue.java
@@ -1,0 +1,6 @@
+package org.seasar.doma.it.criteria;
+
+import org.seasar.doma.DataType;
+
+@DataType
+public record Avenue(String value) {}

--- a/integration-test-java/src/main/java/org/seasar/doma/it/criteria/Place.java
+++ b/integration-test-java/src/main/java/org/seasar/doma/it/criteria/Place.java
@@ -1,0 +1,22 @@
+package org.seasar.doma.it.criteria;
+
+import org.seasar.doma.Entity;
+import org.seasar.doma.Id;
+import org.seasar.doma.Metamodel;
+import org.seasar.doma.Table;
+
+@Entity(metamodel = @Metamodel)
+@Table(name = "address")
+public class Place {
+  @Id private Integer addressId;
+  private Avenue street;
+  private Integer version;
+
+  public Integer getAddressId() {
+    return addressId;
+  }
+
+  public void setAddressId(Integer addressId) {
+    this.addressId = addressId;
+  }
+}

--- a/integration-test-java/src/main/java/org/seasar/doma/it/criteria/Place.java
+++ b/integration-test-java/src/main/java/org/seasar/doma/it/criteria/Place.java
@@ -6,7 +6,7 @@ import org.seasar.doma.Metamodel;
 import org.seasar.doma.Table;
 
 @Entity(metamodel = @Metamodel)
-@Table(name = "address")
+@Table(name = "ADDRESS")
 public class Place {
   @Id private Integer addressId;
   private Avenue street;

--- a/integration-test-java/src/test/java/org/seasar/doma/it/criteria/QueryDslEntityqlSelectTest.java
+++ b/integration-test-java/src/test/java/org/seasar/doma/it/criteria/QueryDslEntityqlSelectTest.java
@@ -420,6 +420,14 @@ public class QueryDslEntityqlSelectTest {
   }
 
   @Test
+  void where_eq_dataType() {
+    Place_ p = new Place_();
+
+    List<Place> list = dsl.from(p).where(c -> c.eq(p.street, new Avenue("STREET 10"))).fetch();
+    assertEquals(1, list.size());
+  }
+
+  @Test
   void innerJoin() {
     Employee_ e = new Employee_();
     Department_ d = new Department_();

